workflow/artifacts/git/git.go

```diff
@@ -28,130 +27,136 @@ type GitArtifactDriver struct {
 	InsecureIgnoreHostKey bool
 }
 
-// Load download artifacts from an git URL
-func (g *GitArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
+func (g *GitArtifactDriver) auth() (func(), transport.AuthMethod, []string, error) {
 	if g.SSHPrivateKey != "" {
 		signer, err := ssh.ParsePrivateKey([]byte(g.SSHPrivateKey))
 		if err != nil {
-			return errors.InternalWrapError(err)
+			return nil, nil, nil, err
+		}
+		privateKeyFile, err := ioutil.TempFile("", "id_rsa.")
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		err = ioutil.WriteFile(privateKeyFile.Name(), []byte(g.SSHPrivateKey), 0600)
+		if err != nil {
+			return nil, nil, nil, err
 		}
 		auth := &ssh2.PublicKeys{User: "git", Signer: signer}
 		if g.InsecureIgnoreHostKey {
 			auth.HostKeyCallback = ssh.InsecureIgnoreHostKey()
 		}
-		return gitClone(path, inputArtifact, auth, g.SSHPrivateKey)
+		args := []string{"ssh", "-i", privateKeyFile.Name()}
+		if g.InsecureIgnoreHostKey {
+			args = append(args, "-o", "StrictHostKeyChecking=no", "-o", "UserKnownHostsFile=/dev/null")
+		} else {
+			args = append(args, "-o", "StrictHostKeyChecking=yes", "-o")
+		}
+		env := []string{"GIT_SSH_COMMAND=" + strings.Join(args, " ")}
+		if g.InsecureIgnoreHostKey {
+			auth.HostKeyCallback = ssh.InsecureIgnoreHostKey()
+			env = append(env, "GIT_SSL_NO_VERIFY=true")
+		}
+		return func() { _ = os.Remove(privateKeyFile.Name()) },
+			auth,
+			env,
+			nil
 	}
 	if g.Username != "" || g.Password != "" {
-		auth := &http.BasicAuth{Username: g.Username, Password: g.Password}
-		return gitClone(path, inputArtifact, auth, "")
+		filename, err := filepath.Abs("git-ask-pass.sh")
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		_, err = os.Stat(filename)
+		if os.IsNotExist(err) {
+			err := ioutil.WriteFile(filename, []byte(`#!/bin/sh
+case "$1" in
+Username*) echo "${GIT_USERNAME}" ;;
+Password*) echo "${GIT_PASSWORD}" ;;
```

Comment:
Possible smell: writing passwords to the file system?

Reply:
this is copied from Argo CD, the actual password is not written to a file